src/app/codegen-data-model-provider/EmberDataBuffer.cpp

```diff
@@ -0,0 +1,358 @@
+CHIP_ERROR OutOfRangeError(unsigned byteCount)
+{
+    if (IsOddIntegerSize(byteCount))
+    {
+        return CHIP_ERROR_INVALID_ARGUMENT;
+    }
+    return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+}
+
+} // namespace
+
+CHIP_ERROR EmberAttributeBuffer::DecodeUnsignedInteger(chip::TLV::TLVReader & reader, EndianWriter & writer)
+{
+    UnsignedDecodeInfo info = GetUnsignedDecodeInfo(mAttributeType);
+
+    // Any size of integer can be read by TLV getting 64-bit integers
+    uint64_t value;
+
+    if (reader.GetType() == TLV::kTLVType_Null)
+    {
+        // we know mIsNullable due to the check at the top of ::Decode
+        value = ~(0ULL); // Null value is ALWAYS negative-fill
+    }
+    else
+    {
+        ReturnErrorOnFailure(reader.Get(value));
+
+        VerifyOrReturnError(mIsNullable // ::max() on the type is used as the NULL flag
+                                ? (value < info.maxValue)
+                                : (value <= info.maxValue),
+                            OutOfRangeError(info.byteCount));
+    }
+
+    writer.EndianPut(value, info.byteCount);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EmberAttributeBuffer::DecodeSignedInteger(chip::TLV::TLVReader & reader, EndianWriter & writer)
+{
+    SignedDecodeInfo info = GetSignedDecodeInfo(mAttributeType);
+
+    // Any size of integer can be read by TLV getting 64-bit integers
+    int64_t value;
+
+    if (reader.GetType() == TLV::kTLVType_Null)
+    {
+        // we know mIsNullable due to the check at the top of ::Decode
+
+        // Most negative integer (i.e. 0b1000...0 is flagged as NULL value)
+        value = info.minValue;
+    }
+    else
+    {
+        ReturnErrorOnFailure(reader.Get(value));
+
+        // NULLABLE reserves minValue for NULL, so range is:
+        //   - NULLABLE:      (minValue, MaxValue]
+        //   - NON-NULLABLE:  [minValue, MaxValue]
+        bool valid = (value <= info.maxValue) && (mIsNullable ? (value > info.minValue) : (value >= info.minValue));
+
+        VerifyOrReturnError(valid, OutOfRangeError(info.byteCount));
+    }
+    writer.EndianPutSigned(value, info.byteCount);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EmberAttributeBuffer::DecodeAsString(chip::TLV::TLVReader & reader, PascalString stringType, TLV::TLVType tlvType,
+                                                EndianWriter & writer)
+{
+    // Handle null first, then the actual data
+    if (reader.GetType() == TLV::kTLVType_Null)
+    {
+        // we know mIsNullable due to the check at the top of ::Decode
+        switch (stringType)
+        {
+        case PascalString::kShort:
+            writer.Put8(0xFF);
+            break;
+        case PascalString::kLong:
+            writer.Put16(0xFFFF);
+            break;
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    const uint32_t stringLength = reader.GetLength();
+
+    VerifyOrReturnError(reader.GetType() == tlvType, CHIP_ERROR_WRONG_TLV_TYPE);
```

Comment:
Shouldn't we do those checks first?

Reply:
The first check is for NULL which also uses GetType. The order seems correct: it is either null (first check) or the expected type.